src/main/java/edu/kit/compiler/register_allocation/ApplyAssignment.java

```diff
@@ -0,0 +1,658 @@
+     *
+     * This applies all steps of the assignment by first handling the
+     * input instructions and adding a function prolog and epilog afterwards.
+     */
+    public static List<String> createFunctionBody(RegisterAssignment[] assignment, RegisterSize[] sizes,
+                                                  Lifetime[] lifetimes, List<Block> ir, int numInstructions,
+                                                  int nArgs, CallingConvention cconv) {
+        ApplyAssignment apply = new ApplyAssignment(
+                assignment, sizes, lifetimes, ir, numInstructions, cconv
+        );
+        AssignmentResult result = apply.doApply();
+        List<String> output = apply.createFunctionProlog(nArgs, result.getUsedRegisters());
+        output.addAll(result.getInstructions());
+        output.addAll(apply.createFunctionEpilog());
+        return output;
+    }
+
+    /**
+     * Creates the whole function body at once.
+     *
+     * This applies all steps of the assignment by first handling the
+     * input instructions and adding a function prolog and epilog afterwards.
+     */
+    public static List<String> createFunctionBody(RegisterAssignment[] assignment, RegisterSize[] sizes,
+                                                  Lifetime[] lifetimes, List<Block> ir, int numInstructions,
+                                                  int nArgs) {
+        return createFunctionBody(assignment, sizes, lifetimes, ir, numInstructions, nArgs, CallingConvention.X86_64);
+    }
+
+    public ApplyAssignment(RegisterAssignment[] assignment, RegisterSize[] sizes,
+                           Lifetime[] lifetimes, List<Block> ir, int numInstructions,
+                           CallingConvention cconv) {
+        assert assignment.length == sizes.length && assignment.length == lifetimes.length;
+        this.assignment = assignment;
+        this.sizes = sizes;
+        this.lifetimes = lifetimes;
+        this.ir = ir;
+        this.cconv = cconv;
+        this.result = new ArrayList<>();
+        this.savedRegisters = Optional.empty();
+        this.numInstructions = numInstructions;
+
+        assertRegistersDontInterfere(assignment, sizes, lifetimes);
+    }
+
+    public ApplyAssignment(RegisterAssignment[] assignment, RegisterSize[] sizes,
+                           Lifetime[] lifetimes, List<Block> ir, int numInstructions) {
+        this(assignment, sizes, lifetimes, ir, numInstructions, CallingConvention.X86_64);
+    }
+
+    /**
+     * Creates concrete instructions for the given input and the necessary information
+     * (used registers) for creating the function prolog and epilog.
+     */
+    public AssignmentResult doApply() {
+        result = new ArrayList<>();
+        LifetimeTracker tracker = new LifetimeTracker();
+        Map<Integer, String> replace = new HashMap<>();
+
+        int i = 0;
+        for (Block b: ir) {
+            output(".L%d:", b.getBlockId());
```

Comment:
This is specific to the ELF file format. If possible, the register allocation should not depend on the output format. (Maybe we could add a generic label instruction that is transformed by the assembly writer?)

Reply:
I don't think we need to support other file formats?

I could of course do some special syntax for block labels instead. Note that in this case we would need to do that for jumps, too. However, would a custom syntax even be better then ELF syntax?

Reply:
We don't need to, but the question is how much we want to hardcode our compiler to certain formats. We could of course use the ELF syntax and transform it to another, when we want to add other file formats. Converting between output formats is not the cleanest solution, but it would suffice for now, I think.

The `Lower` class tests for `Platform.isMac() || Platform.isWindows()`, which is why I thought of it.